File: 0501/code.clj
;;; 因数分解をする関数
(defn pfac-rec [d n acc]
    (cond
    ;; nが1になったらaccを返す
    (= n 1) acc
    ;; nがdで割り切れるならdをaccに追加し、nをn/d、dを2にして再帰する
    (= 0 (rem n d)) (pfac-rec 2 (/ n d) (cons d acc))
    ;; それ以外はdを+1して再帰する
    :else (pfac-rec (inc d) n acc)))

(defn pfac [n] (pfac-rec 2 n '()))

(println (pfac 40))     ; => (5 2 2 2)
(println (pfac 1986))   ; => (331 3 2)
(println (pfac 2026))   ; => (1013 2)

File: 0501/code.py
# Python ではこの処理を再帰で書くと RecursionError になる場合があるため、反復で実装する
def pfac_rec(d, n, acc):
    while True:
        if n == 1:
            return acc
        if n % d == 0:
            acc = [d] + acc
            n = n // d
            d = 2
        else:
            d = d + 1


def pfac(n):
    return pfac_rec(2, n, [])


print(pfac(40))
print(pfac(1986))
print(pfac(2026))
